feat: successor market (#1500)
vegaprotocol/specs: protocol/0028-GOVE-governance.md

* A market usually enters Active status at the proposal's enactment date/time, but some conditions may delay this or cause the market to be Cancelled instead

A proposal to create a market contains
1. a complete market specification as per the [Market Framework](./0001-MKTF-market_framework.md) that describes the market to be created.
1. an enactment time that is at least the *minimum auction duration* after the vote closing time (see [auction spec](./0026-AUCT-auctions.md))

> 1. if the market is meant to be a *sucessor* of a given market then it contains the marketId of the market it's suceeding (parent market) and certain entries in the market proposal must be identical to those of the market it's succeeding. 

Reviewer: Possibly those entries (whatever they end up being) shouldn't even be set in a successor market proposal? Although perhaps there will be different unchangeable fields depending on the market type in future.

Author: Agree. I'll update.

The parent market must be in one of `active` or `suspended` or `trading terminated` states. 
If the parent market is `settled` or `proposed` or `pending` or `cancelled` then the proposal should be rejected at validation stage with an error "parent market cannot be in * state" with * being one of the dis-allowed states above. 
The point of setting up a market to be successor of existing market is to 
a) allow LPs who wish to continue claim their equity-like-share (ELS) by commiting liquidity to the successor market during the pending period and
b) allow the successor market to inherit the insurance pool of the parent market, once the parent market moves from "trading terminated" to "settled" state. 

Note that each market can have exactly one market as a *sucessor* market. 

> - if there already is a market (possibly pending i.e. in opening auction, see [lifecycle spec](./0043-MKTL-market_lifecycle.md)) naming a parent market which is referenced in the proposal then the proposal is rejected with error parent market no longer available. 

Reviewer: Somewhere we probably want the ability to amend a market to remove a child successor (which would also immediately terminate said child I expect)

Author: Yes, that's a good point, if the successor settles or gets cancelled maybe we want to free the parent to have another successor.

Author: Done.

- A market enters a Pending status as soon as the proposal is Successful (before enactment)
- A market usually enters Active status at the proposal's enactment date/time, but some conditions may delay this or cause the market to be Cancelled instead

A proposal to create a market contains

1. a complete market specification as per the [Market Framework](./0001-MKTF-market_framework.md) that describes the market to be created.
1. an enactment time that is at least the _minimum auction duration_ after the vote closing time (see [auction spec](./0026-AUCT-auctions.md))
1. if the market is meant to be a _successor_ of a given market then it contains the marketId of the market it's succeeding (parent market), a parameter called `insurancePoolFraction` which is a decimal in `[0,1]` (i.e. it can be `0` or `1` or anything in between) and certain entries in the market proposal must be identical to those of the market it's succeeding OR the proposal should simply not contain the fields that cannot be changed.

> In particular instrument, settlement asset, margin asset, and `market.value.windowLength` must match.

Reviewer 2: ```suggestion
In a particular instrument, the settlement asset, margin asset, and `market.value.windowLength` must match.
```

> The parent market must be in one of `active` or `suspended` or `trading terminated` states.

Reviewer 3: Allow settled with `successor launch window`.

Reviewer 4: Something else that needs clarifying is when a successor proposal is enacted, and the parent market is still in `active` (or `suspended` due to auction) state: we want to carry over the ELS, but I don't see any reason why it wouldn't be possible for a successor market to be in opening auction (or even leaving the opening auction) while its parent market is in active state. At which point, then, should we transfer the insurance pool fraction, and what should happen with the ELS?

Arguably, LP's would probably prefer it if their ELS transfers over to the successor market when they provide liquidity on that new market for a couple of reasons:

1. Being the first LP to provide stake on the successor market with ELS carry-over is advantageous if other LP's are still on the parent market, being the only LP with ELS from a previous market is a force multiplier.
2. Depending on the window length, this could allow you to keep an eye on `r` for the parent market and with a single transaction, bail if the parent market experiences negative growth, preserving ELS by moving it to a successor market in opening auction (ELS being "frozen" is preferable to ELS shrinking)
3. What if an LP provides liquidity on the successor market, removes stake from the parent, and then (either accidentally or on purpose) creates an LP on the parent market again? Is the ELS shared between parent and successor markets? Does an LP risk losing the ELS they built up over time by accidentally providing stake on the wrong market like this?
4. Based on what @davidsiska-vega said when we discussed this, the ELS is only preserved if the LP provides stake on the new market and _before_ removing stake from the parent market (or waits for their bond to be released once the parent market is settled). That's fine, but we have to be very careful and clear on what happens when stake is no longer provided on either the parent/successor market when both are up and running (e.g. ELS is carried over to the successor, but by providing stake on the parent market after committing to the successor, though a bizarre thing to do just feels incredibly risky - for the LP if we transfer ELS between markets, or very weird and risky WRT fairness if we duplicate/share ELS between parent and successor.

Author: >  I don't see any reason why it wouldn't be possible for a successor market to be in opening auction (or even leaving the opening auction) while its parent market is in active state. At which point, then, should we transfer the insurance pool fraction, and what should happen with the ELS?

1. the party keeps the virtual stake / ELS on the parent market and gets the same virtual stake (if they provided full stake / or virtual stake using the usual delta way).
2. the insurance pool fraction is transferred at that point

Author: @EVODelavega in fact the spec says that the virtual stake is *copied* not *transferred*. This is important.

> If the market has a "parent market" then each LP which commits liquidity (to this market) gets the virtual stake copied from the parent market as the 1st step of the process and the stake they are committing here minus the stake on parent market is treated as the `delta` here.

Reviewer 5: As David says, virtual stake (ELS) is not transferred it is copied.

> very weird and risky WRT fairness if we duplicate/share ELS between parent and successor.
I am not sure what is odd here. Once successor market leaves opening auction, ELS should be _copied_ to the new market, once. It is unchanged on the parent market. 

Where is the fairness issue you see @EVODelavega ?

> The parent market must be in one of `active` or `suspended` or `trading terminated` states.

Author: ```suggestion
The parent market must be in one of `active` or `suspended` or `trading terminated` states or `settled` but with time since settlement less than or equal `network.liquidity.successorLaunchWindowLength`.
```

> The parent market must be in one of `active` or `suspended` or `trading terminated` states or `settled` but with time since settlement less than or equal `network.liquidity.successorLaunchWindowLength`.

Author: ```suggestion
The parent market must be in one of `active` or `suspended` or `trading terminated` states or `settled` but with time since settlement less than or equal `network.liquidity.successorLaunchWindowLength`. In particular the market cannot be `pending`.
```

> The parent market must be in one of `active` or `suspended` or `trading terminated` states or `settled` but with time since settlement less than or equal `network.liquidity.successorLaunchWindowLength`. In particular the market cannot be `pending`.

Author: ```suggestion
The parent market must be in one of `active` or `suspended` or `trading terminated` states or `settled` but with time since settlement less than or equal `network.liquidity.successorLaunchWindowLength` or `closed by governance` but with the closing time less than or equal `network.liquidity.successorLaunchWindowLength`. 
In particular the market cannot be `pending`.
```

> The parent market must be in one of `active` or `suspended` or `trading terminated` states.
> If the parent market is `settled` or `proposed` or `pending` or `cancelled` then the proposal should be rejected at validation stage with an error "parent market cannot be in `*` state" with `*` being one of the dis-allowed states above.
> The point of setting up a market to be successor of existing market is to
> a) allow LPs who wish to continue claim their equity-like-share (ELS) by committing liquidity to the successor market during the pending period and
> b) allow the successor market to inherit the insurance pool of the parent market. When the successor market leaves the opening auction (moves from pending to active) the amount equal to `insurancePoolFraction x parent market insurance pool balance` is transferred to the successor market insurance pool. Once the parent market moves from "trading terminated" to "settled" state the entire remaining insurance pool of the successor market is transferred to the successor market insurance pool.

Reviewer 2: ```suggestion
The parent market must be in one of `active` or `suspended` or `trading terminated` states.
If the parent market is `settled` or `proposed` or `pending` or `cancelled` then the proposal should be rejected at the validation stage with an error "parent market cannot be in `*` state" with `*` being one of the dis-allowed states above.
The point of setting up a market to be successor of an existing market is to
a) allow LPs continue claim their equity-like-share (ELS) by committing liquidity to the successor market during the pending period if they wish to, and
b) allow the successor market to inherit the insurance pool of the parent market. When the successor market leaves the opening auction (moves from pending to active) the amount equal to `insurancePoolFraction x parent market insurance pool balance` is transferred to the successor market insurance pool. Once the parent market moves from "trading terminated" to "settled" state, the entire remaining insurance pool of the successor market is transferred to the successor market insurance pool.
```

Note that each market can have exactly one market as a _successor_ market.

> - if there already is a market (possibly pending i.e. in opening auction, see [lifecycle spec](./0043-MKTL-market_lifecycle.md)) naming a parent market which is referenced in the proposal then the proposal is rejected with error parent market no longer available.
> - if there are two proposals naming the same parent market then whichever one gets into the pending state first (i.e. passes governance vote) becomes the successor of the named parent; the other proposal is cancelled with reason "parent market no longer available".
> - if there is a successor market naming a parent market that terminates and settles or is cancelled by governance before the parent market (for whatever reason) then the parent market successor is set to empty and a different successor can be proposed.

Reviewer 2: ```suggestion
- if there already is a market (possibly pending, i.e. in opening auction, see [lifecycle spec](./0043-MKTL-market_lifecycle.md)), naming a parent market, then a subsequent proposal referencing that market is rejected.
- if there are two proposals naming the same parent market then whichever one gets into the pending state first (i.e. passes governance vote) becomes the successor of the named parent; the other proposal is cancelled with reason "parent market no longer available".
- if there is a successor market naming a parent market that terminates and settles or is cancelled by governance before the parent market (for whatever reason) then the parent market successor is set to empty and a different successor can be proposed.
```
can the updated successor market be proposed using an "update market" proposal? Or is it some other action? Does that make that field conditionally immutable?
